### PR TITLE
Rich text formatting filter

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

See also #4658.

Related PRs: #9192.

This small change makes the formatting controls for all rich text instances extensible. E.g. a plugin wants to add an extra button to add a "code" tag:

```js
addFilter( 'editor.richText.formattingControls', 'core/test', ( controls ) => {
	return [
		...controls,
		{
			icon: 'editor-code',
			title: __( 'Code' ),
			shortcut: displayShortcut.primary( 'x' ),
			format: 'code',
		},
	];
} );
```

For the moment it's limited to adding formats, but I could imagine an option to insert custom HTML as well. Let's start simple. :)

Questions:

* What's the right place to call `applyFilters`? At the top level, plugins will need to add the filter before the Gutenberg scripts run. At render it will be run too many times. Should it be run on constructing `FormatToolbar`?
* If we keep `DEFAULT_CONTROLS`, we this might need a filter too, but it does not make sense without context. `RichText` is not aware of the block using it. It could also be the tag, but not sure if this makes any sense. Alternatively we could get rid of `DEFAULT_CONTROLS` or invert the behaviour, e.g. a button would declare controls to omit rather then the ones to display.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation.

### CODE REVIEW

**Comment by gziolo** [comment]
> * What's the right place to call `applyFilters`? At the top level, plugins will need to add the filter before the Gutenberg scripts run. At render it will be run too many times. Should it be run on constructing `FormatToolbar`?

Ideally, at any time to make sure it also works in the case when plugin code is executed after an editor has intialized :) Initial implementation can take a simpler assumption and apply filters when formatting toolbar renders for the first time. The easier way to implement it is to convert `FORMATTING_CONTROLS` into a function and wrap it with `once` utility from lodash. This way you ensure all the filters are computed only once.

> * If we keep `DEFAULT_CONTROLS`, we this might need a filter too, but it does not make sense without context. `RichText` is not aware of the block using it. It could also be the tag, but not sure if this makes any sense. Alternatively we could get rid of `DEFAULT_CONTROLS` or invert the behaviour, e.g. a button would declare controls to omit rather then the ones to display.

Can we assume that when `enabledControls` are not provided, all controls are allowed? It seems to be the case at the moment, anyway.

File: editor/components/rich-text/format-toolbar/index.js [hunk 1]
```diff
@@ -22,7 +23,7 @@ import { filterURLForDisplay } from '../../../utils/url';
 
 const { ESCAPE, LEFT, RIGHT, UP, DOWN, BACKSPACE, ENTER, displayShortcut } = keycodes;
 
-const FORMATTING_CONTROLS = [
+const FORMATTING_CONTROLS = applyFilters( 'editor.richText.formattingControls', [
```
Comment (gziolo) [inline editor/components/rich-text/format-toolbar/index.js]: For consistency, we should name this filter `editor.RichText.formattingControls`.

**Comment by youknowriad** [comment]
How can we make this tweakable per block? I want to add a button to the Paragraph and Heading but not the captions of images etc...

**Comment by gziolo** [comment]
You can also check how autocompleters were made extensible by @brandonpayton using a higher-order component in here: https://github.com/WordPress/gutenberg/blob/master/editor/components/autocomplete/index.js. It is much more complex use case, because every autocompleter needs to keep its state independently.

> How can we make this tweakable per block? I want to add a button to the Paragraph and Heading but not the captions of images etc...

`BlockEditContext` is your friend, `blocks.Autocomplete.completers` filter takes `blockName` as 2nd param :)

**Comment by grappler** [comment]
Where are the formats defined that could be used?

**Comment by iseulde** [comment]
> How can we make this tweakable per block? I want to add a button to the Paragraph and Heading but not the captions of images etc...

I'm not sure about this... There will be blocks with multiple RichText fields, so what if you only want them enabled on one field? Maybe we should allow two ways of registering control?

* You can register controls globally for RichText with the filter. This will add them to all RichText instances.
* You can pass additional controls to the RichText instance within the block registration.

Alternatively we could pass the block name _and_ the tag name as arguments on the filter.

**Comment by gziolo** [comment]
Yes, the more params we provide the better control plugin developers will get. I like the idea of passing a tag name 👍

**Comment by iseulde** [comment]
> Where are the formats defined that could be used?

They are defined here: https://github.com/tinymce/tinymce/blob/10534f7dead4bb79bf8a98d91c2465e1ee652b66/src/core/main/ts/fmt/DefaultFormats.ts

But this is very limited. Eventually I think it should be possible to pass a function to call on the selection, and return a React element.

**Comment by gziolo** [comment]
Rebased with master. I think we will need two filters:
- `editor.RichText.formattingControls` - which will allow to customize the list of the existing formatting controls (add/update/remove)
- `editor.RichtText.enabledFormattingControls` - which will allow to customize set of controls per block type, the idea is that to render control it would have to be registered as `formattingControl` and present in this array.

I'm AFK for the rest of the week so won't have chance to work on it until Monday.

**Comment by noisysocks** [comment]
While working on https://github.com/WordPress/gutenberg/pull/8807 I was thinking how nice it would be if our _Link_ button was entirely implemented using the `formattingControls` API that we expose to third parties.

I would especially love if the API were flexible enough that we could remove [our  complex link formatting logic](https://github.com/WordPress/gutenberg/blob/7b73a2067f5f1ecd387a4631be1290a040ddcb14/packages/editor/src/components/rich-text/index.js#L828-L901) out of `RichText` and into the `link` `formattingControls` object.

Rough and not-very-thought-out pseudocode of what I'm getting at:

```js
const plugin = {
	icon: 'admin-links',
	title: __( 'Link' ),
	shortcut: displayShortcut.primary( 'k' ),
	format: 'link',
	attributes: {
		href: { type: 'string', source: 'attribute', attribute: 'href', default: null },
	},
	renderAdditionalUI( format, setFormat ) {
		return <LinkContainer format={ format } setFormat={ setFormat } />;
	},
	apply( attributes, editor ) {
		if ( attributes.href ) {
			if ( editor.getSelection().isCollapsed() ) {
				editor.insertContent( editor.dom.createHTML(
					'a',
					{ href: attributes.href },
					editor.dom.encode( attributes.href )
				) );
			} else {
				editor.execCommand( 'mceInsertLink', {
					href: attributes.href,
					'data-wp-placeholder': null,
					'data-mce-bogus': null,
				} );
			}
		} else {
			editor.formatter.apply( 'link', {
				href: '#',
				'data-wp-placeholder': true,
				'data-mce-bogus': true,
			} );
		}
	},
	remove( format, editor ) {
		editor.execCommand( 'Unlink' );
	},
},
```

**Comment by jcklpe** [comment]
ooooh, sorry to chime in. I'm just following along through all the github issues on this subject, reading how you guys are solving the issue and this is really clever. I mean I'm still a novice at code but this solution sounds really nice and flexible. Keep up the good work guys!

**Comment by iseulde** [comment]
I'm going to close this as I think #10068 is the way forward. Feel free to comment with thoughts on it. This may address some concerns @noisysocks has regarding the flexibility of the API.

**Pull request closed without merging**